### PR TITLE
BigMac Alerting Rules

### PR DESCRIPTION
**Pull request by ssyno**
Before adding a new alerting rule into this repository you should consider creating an SLO rules instead.
SLO helps you both increase the quality of your monitoring and reduce the alert noise.

* How to create a SLO rule: https://github.com/giantswarm/sloth-rules#how-to-create-a-slo
* Documentation: https://intranet.giantswarm.io/docs/monitoring/slo-alerting/

---
Towards: https://github.com/giantswarm/roadmap/issues/3085 & https://github.com/giantswarm/roadmap/issues/3094
This PR adds basic alerting on BigMac's apps along with some specific teleport-operator ones.

### Checklist

- [x] Update CHANGELOG.md
- [x] Add [Unit tests](https://github.com/giantswarm/prometheus-rules/#testing)
- [x] Follow [Alert structure](https://github.com/giantswarm/prometheus-rules/#how-alerts-are-structured)
- [ ] Consider [creating a dashboard](https://docs.giantswarm.io/getting-started/observability/grafana/custom-dashboards/) ([guidelines](https://intranet.giantswarm.io/docs/product/ux/guidelines/dashboards/)) (if it does not exist already) to help oncallers monitor the status of the issue.
- [x] Request review from oncall area, as well as team (e.g: `oncall-kaas-cloud` GitHub group).

### CODE REVIEW

**Review by tuladhar — APPROVED** [review]

File: helm/prometheus-rules/templates/alerting-rules/teleport.rules.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      for: 30m
+      labels:
+        area: kaas
+        cancel_if_cluster_status_creating: "true"
+        cancel_if_cluster_status_updating: "true"
+        cancel_if_cluster_status_deleting: "true"
+        cancel_if_outside_working_hours: "true"
+        severity: page
+        team: bigmac
+        topic: teleport
```
Comment (QuantumEnigmaa) [inline helm/prometheus-rules/templates/alerting-rules/teleport.rules.yaml:40]:
```suggestion
        topic: teleport

```

File: helm/prometheus-rules/templates/alerting-rules/teleport.rules.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  creationTimestamp: null
+  labels:
+    {{- include "labels.common" . | nindent 4 }}
+  name: prometheus.rules
```
Comment (hervenicol) [inline helm/prometheus-rules/templates/alerting-rules/teleport.rules.yaml:7]: This name should be unique.
```suggestion
  name: teleport.rules
```

**Review by hervenicol — COMMENTED** [review]
I'm a bit sad there's no unit test, especially for the new teleport alerts.


If you have trouble understanding how these work, ping me and we can pair on your first prometheus rule UT!

These rules are quite simple and easy to understand, but UTs can serve as examples of what data we expect, and help understand corner cases or strange behaviours as well later.

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Added specific deployment status alerts for BigMac application
+- Added telepor-operator alerts for missing secrets and configmaps
```
Comment (marieroque) [inline CHANGELOG.md:13]:
```suggestion
- Added teleport-operator alerts for missing secrets and configmaps
```

**Comment by ssyno** [comment]
> I'm a bit sad there's no unit test, especially for the new teleport alerts.  [ 68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f424c4a7932783651777a6764724366416c442f67697068792e676966 ](https://camo.githubusercontent.com/6394767b6c9dbae1dfc2c81d9876a3cc1580723437dc9d4a815a540235060f1c/68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f424c4a7932783651777a6764724366416c442f67697068792e676966) [ ](https://camo.githubusercontent.com/6394767b6c9dbae1dfc2c81d9876a3cc1580723437dc9d4a815a540235060f1c/68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f424c4a7932783651777a6764724366416c442f67697068792e676966)
> 
> If you have trouble understanding how these work, ping me and we can pair on your first prometheus rule UT!
> 
> These rules are quite simple and easy to understand, but UTs can serve as examples of what data we expect, and help understand corner cases or strange behaviours as well later.

hey @hervenicol,
Thanks for the guidance!
I have also successfully added unit testing for those 2 new teleport alerts.

File: test/tests/providers/capi/capz/teleport.rules.test.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  - interval: 1m
+    input_series:
+      - series: 'kube_configmap_info{app="kube-state-metrics", cluster_id="grizzly", configmap="grizzly-teleport-kube-agent-config"}'
+        values: "1+0x40"
+      - series: 'capi_cluster_info{cluster_id="grizzly", control_plane_reference_kind="KubeadmControlPlane"}'
+        values: "1+0x40"
+    alert_rule_test:
+      - alertname: TeleportKubeAgentConfigMapMismatch
+        eval_time: 30m
+        exp_alerts: []
```
Comment (hervenicol) [inline test/tests/providers/capi/capz/teleport.rules.test.yml:73]: You could merge test cases in one set of test series, like this:

```suggestion
  - interval: 1m
    input_series:
      - series: 'kube_configmap_info{app="kube-state-metrics", cluster_id="grizzly", configmap="grizzly-teleport-kube-agent-config"}'
        values: "1+0x40 1+0x40"
      - series: 'kube_configmap_info{app="kube-state-metrics", cluster_id="grizzly", configmap="test-teleport-kube-agent-config"}'
        values: "_x40   1+0x40"
      - series: 'capi_cluster_info{cluster_id="grizzly", control_plane_reference_kind="KubeadmControlPlane"}'
        values: "1+0x40 1+0x40"
    alert_rule_test:
      - alertname: TeleportKubeAgentConfigMapMismatch
        eval_time: 30m
        exp_alerts: []
      - alertname: TeleportKubeAgentConfigMapMismatch
        eval_time: 70m
        exp_alerts:
          - exp_labels:
              alertname: TeleportKubeAgentConfigMapMismatch
              area: kaas
              cancel_if_cluster_status_creating: "true"
              cancel_if_cluster_status_updating: "true"
              cancel_if_cluster_status_deleting: "true"
              cancel_if_outside_working_hours: "true"
              severity: page
              team: bigmac
              topic: teleport
            exp_annotations:
              description: "Mismatch in number of teleport-kube-agent-config secrets and clusters"
```

The `suggestion` does not show it clearly, so basically I did:
- concatenate values for the series
- have first a test where we have no alert (at 30m)
- then a test with an alert (40m later, at 70m)

This way you don't have to duplicate your series and all their labels. Also, I think it shows better how series could change over time.

**Review by hervenicol — APPROVED** [review]
Now I like it!
